src/rw.jl

```diff
@@ -0,0 +1,196 @@
+    """
+    function write_setup(variables ...; bp_filename = "")
+
+        io = ADIOS2.declare_io(adios, "IO")
+
+        for var_pair in variables
+            var_id = define_variable(io, var_pair.first, eltype(var_pair.second))  # Define a new variable
+            push!(vars, var_id)
+        end
+
+        bp_path = joinpath(pwd(), bp_filename)
+        global engine = ADIOS2.open(io, bp_path, mode_write)   # Open the file/stream from the .bp file
+
+    end
+
+    """
+    Perform the update using specified variables.
+    """
+    function write(update_var = nothing ...)
+
+        for v in vars
+            begin_step(engine)                                       # Begin ADIOS2 write step
+            put!(engine, v, update_var)                              # Add update_var to variables for writing
+            end_step(engine)                                         # End ADIOS2 write step (normally, also includes the actual writing of data)
+        end
+
+    end
+
+    """
+    Encodes the loop condition for the reading step.
+    """
+
+    function reading(; timeout = 100.0)
+
+        loop_condition = begin_step(engine, step_mode_read, timeout) != step_status_end_of_stream
+
+        if loop_condition
+            global reading_now = true
+            return true
+        else
+            global reading_now = false
+            global nprocessed = 0
+            return false
+        end
+        
+    end
+
+    """
+    Reads the data sequentially. This function is particularly useful when called from a loop.
+    """
+
+    function read(var; verbose = true)
+
+        var_id = inquire_variable(io, var)
+
+        if nprocessed == 0
+            nxy_global = shape(var_id)                                               # Extract meta data
+            nxy        = count(var_id)                                               # ...
+            var_type   = type(var_id)                                                # ...
+            global V   = zeros(var_type, nxy)                                        # Preallocate memory for the variable using the meta data
+        end
+
+        get(engine, var_id, V)
+
+        if reading                                                                  # end the step automatically if loop detected
+            end_step(engine)
+            global nprocessed += 1
+        end
+
+        if verbose
+            print("Variable: " * string(var))
+            print("Step: " * string(nprocessed))
+            print("Variable steps: " * string(steps(var_iq)))
+            print("Current step: " * string(current_step(engine)))
+        end
+
+    end
+
+    """
+    Provide details about all avaiable variables listed in the ADIOS2 file.
+    """
+
+    function inspect_variables(filename::AbstractString = "")
+        if !isempty(filename)
+            var_info = adios_all_variable_names(filename)
+            print(var_info)
+        else
+            print("error")
```

Comment:
Use exceptions defined in the file Exceptions.jl. Add more exception types there if appropriate. See here for examples: https://github.com/omlins/ParallelStencil.jl/blob/main/src/ParallelKernel/Exceptions.jl